runtime-modules/content-directory/src/lib.rs

```diff
@@ -1689,88 +1858,311 @@ impl<T: Trait> Module<T> {
+            Some(entity_property_values_updated)
+        } else {
+            None
+        }
     }
 
-    /// Get all referenced entities from corresponding property with `SameOwner` flag set,
-    /// call `retrieve_all_entities_to_perform_ownership_transfer` recursively to complete tree traversal
-    pub fn get_all_same_owner_entities(
-        class: &Class<T>,
-        value: &PropertyValue<T>,
-        entities: &mut BTreeSet<T::EntityId>,
-    ) {
-        if let Some(entity_ids) = value.get_involved_entities() {
-            entity_ids.into_iter().for_each(|entity_id| {
-                // If new entity with `SameOwner` flag set found
-                if !entities.contains(&entity_id) {
-                    entities.insert(entity_id);
-                    let new_entity = Self::entity_by_id(entity_id);
-                    Self::retrieve_all_entities_to_perform_ownership_transfer(
-                        &class, new_entity, entities,
-                    );
-                }
+    /// Perform all checks to ensure `property_values` under provided `property_ids` are valid
+    pub fn ensure_property_values_are_valid(
+        class_properties: &[Property<T>],
+        property_ids: &BTreeSet<PropertyId>,
+        entity_controller: &EntityController<T>,
+        property_values: &BTreeMap<PropertyId, PropertyValue<T>>,
+    ) -> dispatch::Result {
+        for property_id in property_ids {
+            // Indexing is safe, Class should always maintain such constistency
+            let class_property = &class_properties[*property_id as usize];
+
+            if let Some(new_value) = property_values.get(property_id) {
+                // Validate new PropertyValue against the type of this Property and check any additional constraints
+                class_property
+                    .ensure_property_value_to_update_is_valid(new_value, entity_controller)?;
+            } else {
+                // All required property values should be provided
+                ensure!(!class_property.required, ERROR_MISSING_REQUIRED_PROP);
+            }
+        }
+        Ok(())
+    }
+
+    /// Ensure all `updated_values_for_existing_properties` satisfy unique option, if required
+    pub fn ensure_property_values_unique_option_satisfied(
+        updated_values_for_existing_properties: ValuesForExistingProperties<T>,
+    ) -> dispatch::Result {
+        for updated_value_for_existing_property in updated_values_for_existing_properties.values() {
+            let (property, value) = updated_value_for_existing_property.unzip();
+
+            // Ensure all PropertyValue's with unique option set are unique, except of null non required ones
+            property
+                .ensure_unique_option_satisfied(value, &updated_values_for_existing_properties)?;
+        }
+        Ok(())
+    }
+
+    /// Perform all necessary checks to ensure `new_values_for_existing_properties` are valid
+    pub fn ensure_new_property_values_are_valid(
```

Comment:
Is this needed, what is the meaning of it vs. `ensure_property_values_are_valid`.
Please reply only after having done refactor on that method first.

Reply:
Thanks for pointing this issue. Made a reusable version of `ensure_property_values_are_valid` here https://github.com/Joystream/joystream/pull/669/commits/affc8b4b9b482df85445f3c3ba924fd586f42dee